document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/transformer/PostgresUnnestQueryTransformer.java

```diff
@@ -0,0 +1,267 @@
+    return fromTypeExpressions.stream()
+        .map(
+            fromTypeExpression ->
+                (UnnestExpression) fromTypeExpression.accept(unnestExpressionProvider))
+        .filter(
+            unnestExpression ->
+                filterTypeExpression.accept(
+                    new FilterToUnnestExpressionMatcher(
+                        filterIdentifierProvider, unnestExpression)))
+        .max(Comparator.comparingInt(u -> u.getIdentifierExpression().getName().length()));
+  }
+
+  private static UnnestExpression getBuildUnnestExpression(
+      UnnestExpression matchedUnnestExpression, FilterTypeExpression filterTypeExpression) {
+    return UnnestExpression.builder()
+        .identifierExpression(matchedUnnestExpression.getIdentifierExpression())
+        .preserveNullAndEmptyArrays(matchedUnnestExpression.isPreserveNullAndEmptyArrays())
+        .filterTypeExpression(
+            buildAndFilter(filterTypeExpression, matchedUnnestExpression.getFilterTypeExpression()))
+        .build();
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class UnnestExpressionProvider implements FromTypeExpressionVisitor {
+
+    @Override
+    public UnnestExpression visit(UnnestExpression unnestExpression) {
+      return unnestExpression;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class FilterIdentifierProvider implements SelectTypeExpressionVisitor {
+
+    @Override
+    public List<String> visit(AggregateExpression expression) {
+      return expression.getExpression().accept(this);
+    }
+
+    @Override
+    public List<String> visit(ConstantExpression expression) {
+      return List.of();
+    }
+
+    @Override
+    public List<String> visit(FunctionExpression expression) {
+      return expression.getOperands().stream()
+          .map(exp -> exp.<List<String>>accept(this))
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<String> visit(IdentifierExpression expression) {
+      return List.of(expression.getName());
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static class AndFiltersCollector implements FilterTypeExpressionVisitor {
+
+    @Override
+    public List<FilterTypeExpression> visit(LogicalExpression expression) {
+      if (expression.getOperator().equals(LogicalOperator.AND)) {
+        return expression.getOperands().stream()
+            .map(
+                filterTypeExpression ->
+                    filterTypeExpression.<List<FilterTypeExpression>>accept(this))
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+      }
+      return List.of(expression);
+    }
+
+    @Override
+    public List<FilterTypeExpression> visit(RelationalExpression expression) {
+      return List.of(expression);
+    }
+
+    @Override
+    public List<FilterTypeExpression> visit(KeyExpression expression) {
+      return List.of(expression);
```

Comment:
Do we really need to collect `KeyExpression`s?
I think it is safe to skip them?
On the other hand, I'm thinking about queries of the form `SELECT * FROM collection, UNNEST ... WHERE id=? OR sales.medium.type=?`

In this case, the whole filter (including the id filter) goes inside the unnest filter?

Reply:
No. Fixed it.